server/schedulers/evict_slow_trend.go

```diff
@@ -0,0 +1,437 @@
+	err := s.prepareEvictLeader(cluster, s.conf.popCandidate())
+	if err != nil {
+		log.Info("prepare for evicting leader by slow trend failed", zap.Error(err), zap.Uint64("store-id", slowStoreID))
+		storeSlowTrendActionStatusGauge.WithLabelValues("evict.prepare.err").Inc()
+		return ops, nil
+	}
+	storeSlowTrendActionStatusGauge.WithLabelValues("evict.start").Inc()
+	return s.scheduleEvictLeader(cluster), nil
+}
+
+func newEvictSlowTrendScheduler(opController *schedule.OperatorController, conf *evictSlowTrendSchedulerConfig) schedule.Scheduler {
+	return &evictSlowTrendScheduler{
+		BaseScheduler: NewBaseScheduler(opController),
+		conf:          conf,
+	}
+}
+
+func chooseEvictCandidate(cluster schedule.Cluster) (slowStore *core.StoreInfo) {
+	stores := cluster.GetStores()
+	if len(stores) <= 3 {
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.none:too-few").Inc()
+		return
+	}
+	var candidates []*core.StoreInfo
+	for _, store := range stores {
+		if store.IsRemoved() {
+			continue
+		}
+		if !(store.IsPreparing() || store.IsServing()) {
+			continue
+		}
+		slowTrend := store.GetSlowTrend()
+		if slowTrend != nil && slowTrend.CauseRate > 0 {
+			if slowTrend.ResultRate < 0 {
+				candidates = append(candidates, store)
+				storeSlowTrendActionStatusGauge.WithLabelValues("cand.add").Inc()
+			}
+		}
+	}
+	if len(candidates) == 0 {
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.none:no-fit").Inc()
+		return
+	}
+
+	// TODO: Calculate to judge if one store is way slower than the others
+	if len(candidates) > 1 {
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.none:too-many").Inc()
+		return
+	}
+
+	store := candidates[0]
+	if !checkStoreSlowerThanOthers(cluster, store) {
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.none:not-slower").Inc()
+		return
+	}
+	return store
+}
+
+func checkStoresAreUpdated(cluster schedule.Cluster, baseline *core.StoreInfo) bool {
+	stores := cluster.GetStores()
+	if len(stores) <= 3 {
+		return false
+	}
+	expected := (len(stores) + 1) / 2
+	updatedStores := 0
+	baselineTs := baseline.GetLastHeartbeatTS()
+	for _, store := range stores {
+		if store.IsRemoved() {
+			updatedStores += 1
+			continue
+		}
+		if !(store.IsPreparing() || store.IsServing()) {
+			updatedStores += 1
+			continue
+		}
+		if store.GetID() == baseline.GetID() {
+			updatedStores += 1
+			continue
+		}
+		if baselineTs.Before(store.GetLastHeartbeatTS()) {
+			updatedStores += 1
+		}
+	}
+	storeSlowTrendMiscGauge.WithLabelValues("stores.check-updated:count").Set(float64(updatedStores))
+	storeSlowTrendMiscGauge.WithLabelValues("stores.check-updated:expected").Set(float64(expected))
+	return updatedStores >= expected
+}
+
+func checkStoreSlowerThanOthers(cluster schedule.Cluster, target *core.StoreInfo) bool {
+	stores := cluster.GetStores()
+	expected := (len(stores)*2 + 1) / 3
+	if expected >= len(stores) {
+		expected = len(stores) - 1
+	}
+	if expected < 1 {
+		expected = 1
+	}
```

Comment:
Maybe here can be tidied as follows:
```Go
if expected < 1 {
	expected = 1
} else if expected >= len(stores) {
	expected = len(stores) - 1
}
	
```

Reply:
It's not the same, the original can handle `len(stores) == 1`

Reply:
It's redundant, removed